pkg/election/leadership.go

```diff
@@ -182,26 +188,81 @@ func (ls *Leadership) Watch(serverCtx context.Context, revision int64) {
 	if ls == nil {
 		return
 	}
+
+	interval := detectHealthyInterval
+	unhealthyTimeout := watchLoopUnhealthyTimeout
+	failpoint.Inject("fastTick", func() {
+		unhealthyTimeout = 5 * time.Second
+		interval = 1 * time.Second
+	})
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	lastHealthyTime := time.Now()
+
 	watcher := clientv3.NewWatcher(ls.client)
 	defer watcher.Close()
-	ctx, cancel := context.WithCancel(serverCtx)
-	defer cancel()
-	// The revision is the revision of last modification on this key.
-	// If the revision is compacted, will meet required revision has been compacted error.
-	// In this case, use the compact revision to re-watch the key.
+	var watchChanCancel *context.CancelFunc
+	defer func() {
+		if watchChanCancel != nil {
+			(*watchChanCancel)()
+		}
+	}()
 	for {
 		failpoint.Inject("delayWatcher", nil)
-		rch := watcher.Watch(ctx, ls.leaderKey, clientv3.WithRev(revision))
-		for wresp := range rch {
+		if watchChanCancel != nil {
+			(*watchChanCancel)()
+		}
+		// In order to prevent a watch stream being stuck in a partitioned node,
+		// make sure to wrap context with "WithRequireLeader".
+		watchChanCtx, cancel := context.WithCancel(clientv3.WithRequireLeader(serverCtx))
+		watchChanCancel = &cancel
+
+		// When etcd is not available, the watcher.Watch will block,
+		// so we check the etcd availability first.
+		if !etcdutil.IsHealthy(serverCtx, ls.client) {
+			if time.Since(lastHealthyTime) > unhealthyTimeout {
+				log.Error("the connect of leadership watcher is unhealthy",
+					zap.Int64("revision", revision),
+					zap.String("leader-key", ls.leaderKey),
+					zap.String("purpose", ls.purpose))
+				return
+			}
+			select {
+			case <-serverCtx.Done():
+				// server closed, return
+				return
+			case <-ticker.C:
+				// continue to check the etcd availability
+				continue
+			}
+		}
+
+		watchChan := watcher.Watch(watchChanCtx, ls.leaderKey, clientv3.WithRev(revision))
```

Comment:
Regarding #6554 "_TSO primary election took 14 minutes after PD(API) Pods being deleted at the same time and PD(API) leader being re-elected_",  I have two questions:
1. If watch can't exit when etcd isn't available is the main reason, it only explains the seconday, which watches the leader here, can't be re-elected as the new primary, and it can't explain why the old primary can't be re-elected as the new primary. What happened on the old primary? Did the old primary also get stuck in some watch loop?
2. It seems to me what we're thinking is, during the long duration between PD(API) Pods being deleted at the same time and PD(API) leader being re-elected, the watch here won't exit, and even after PD (API) leader is back, the watch here somehow still can't catch the lease lost/expired event. Is my understanding correct? If yes, how long is the unsafe duration of PD(API) leader being offline (once this unsafe duration is exceeded, the watcher here still can't catch the lease lost/expired event)? And the setting value of "unhealthyTimeout" needs to be less than this unsafe duration.

Reply:
I do think the current PR can help to mitigate the situation even without perfect solution or answers to 1 and 2, because with this pr, it seems that the secondary can be elected as the new primary which addressed the question 1 above and the question 2 is a parameter tuning problem.

Reply:
It's not sure because the watch loop can't exit that it takes fifteen minutes to re-elect, there could be some reasons, stuck on [grpc retry](https://github.com/tikv/pd/pull/6577#discussion_r1259144016), stuck on [tcp retry](https://github.com/tikv/pd/issues/6554#issuecomment-1582790457), or watch stuck on etcd watch.